File: backend/vortex-admin-pro/src/main/java/com/vortexadmin/service/TwoFactorService.java
package com.vortexadmin.service;

import com.vortexadmin.dto.response.TwoFactorSetupResponse;
import com.vortexadmin.dto.response.TwoFactorStatusResponse;

/**
 * Service contract for TOTP-based two-factor authentication (2FA) lifecycle management,
 * including status retrieval, setup initiation, code verification, and deactivation.
 */
public interface TwoFactorService {

    /**
     * Returns the current 2FA enrollment status for the currently authenticated user.
     *
     * @return a {@link TwoFactorStatusResponse} indicating whether 2FA is enabled and,
     *         if so, the enrollment details
     */
    TwoFactorStatusResponse getStatus();

    /**
     * Initiates the 2FA setup process for the currently authenticated user by generating a
     * TOTP secret and returning a QR code URL and the raw secret for manual entry.
     *
     * @return a {@link TwoFactorSetupResponse} containing the OTP auth URI and backup secret
     */
    TwoFactorSetupResponse setup();

    /**
     * Verifies the provided TOTP code against the pending setup secret and, if valid, activates
     * 2FA for the currently authenticated user.
     *
     * @param code the 6-digit TOTP code from the authenticator application
     * @return a {@link TwoFactorStatusResponse} reflecting the newly enabled 2FA state
     * @throws com.vortexadmin.exception.ApiException if the code is invalid or the setup has
     *         not been initiated
     */
    TwoFactorStatusResponse verifyAndEnable(String code);

    /**
     * Disables 2FA for the currently authenticated user after verifying the provided TOTP code.
     *
     * @param code the 6-digit TOTP code used to confirm the disable action
     * @throws com.vortexadmin.exception.ApiException if the code is invalid or 2FA is not
     *         currently enabled for the user
     */
    void disable(String code);
}

File: frontend/src/services/twoFactorService.js
import api from '../api/axios';

/**
 * Service object for managing TOTP-based two-factor authentication.
 */
export const twoFactorService = {
    /**
     * Retrieves the current 2FA enablement status for the authenticated user.
     * GET /2fa/status
     * @returns {Promise<import('axios').AxiosResponse>} Response containing `{ enabled, remainingBackupCodes }`.
     */
    getStatus: () => api.get('/2fa/status'),

    /**
     * Initiates the 2FA setup flow, returning a TOTP secret and OTP auth URL
     * for QR code generation.
     * POST /2fa/setup
     * @returns {Promise<import('axios').AxiosResponse>} Response containing `{ secret, otpAuthUrl }`.
     */
    setup: () => api.post('/2fa/setup'),

    /**
     * Verifies the TOTP code entered by the user to activate 2FA and receive
     * one-time backup codes.
     * POST /2fa/verify
     * @param {string} code - The 6-digit TOTP code from the authenticator app.
     * @returns {Promise<import('axios').AxiosResponse>} Response containing `{ backupCodes }`.
     */
    verify: (code) => api.post('/2fa/verify', { code }),

    /**
     * Disables 2FA for the authenticated user after verifying the current TOTP
     * or a backup code.
     * POST /2fa/disable
     * @param {string} code - A valid TOTP code or backup code.
     * @returns {Promise<import('axios').AxiosResponse>}
     */
    disable: (code) => api.post('/2fa/disable', { code }),
};
